### DISCUSSION TITLE
Conflict marker with unified format.

### DISCUSSION THREAD

**Discussion by black7375** [jj-vcs/jj#1871] [Ideas]
I had forgotten after posting the following proposal from git-branchless.
- https://github.com/arxanas/git-branchless/discussions/923

I recently remind it.
I'm not sure if it's a valid proposal in this project, but I'm curious about your opinion.

## 1. Snapshot and diff
My opinion is to use snapshot if there are many changes and diff if there are few.
Also, for diff, let's mark it as [Unified Format](https://www.gnu.org/software/diffutils/manual/html_node/Unified-Format.html).
I think it's more consistent, although it's slightly changed from the original proposal.

Diff and Snapshot
```
  <<<<<<<
  @@ -1,4 +1,4 @@ Start diff base-ours
   apple
  -grape
  +grapefruit
   orange
  ======= # Start Snapshot theirs
  APPLE
  GRAPE
  ORANGE
  >>>>>>>
```

Snapshot and Diff
```
  <<<<<<< ours
  APPLE
  GRAPE
  ORANGE
  =======
  @@ -1,4 +1,4 @@ Start diff base-theirs
   apple
  -grape
  +grapefruit
   orange
  >>>>>>>
```

Diff and Diff style
```
  <<<<<<<
  @@ -1,4 +1,4 @@ Start diff base-ours
   apple
  -grape
  +grapefruit
   orange
  =======
  @@ -1,4 +1,5 @@ Start diff base-theirs
  -apple
  +banana
   grape
   orange
  +tomato
  >>>>>>>
```

## 2. Binary
In the case of a conflict of the binary, I personally use a [simple tool](https://github.com/black7375/binary-conflict) to generate `base` and `their` files.

![image](https://user-images.githubusercontent.com/25581533/235336209-b818da65-9a7c-4d80-af6f-c468c856e216.png)

**Comment by martinvonz** [comment]
I don't think the headers from the unified diff format add much. The line numbers are not needed for finding the right location because the hunk is already in the right place, and the labels don't provide any information that isn't easy to find by looking above the conflict hunk.

Regarding binary files, you should be able to [configure your tool as a custom merge tool for `jj resolve`](https://github.com/martinvonz/jj/blob/main/docs/config.md#3-way-merge-tools-for-conflict-resolution).

> My opinion is to use snapshot if there are many changes and diff if there are few.

That might make sense. My main concern is that it implies a seemingly arbitrary threshold. #823 is also related.

**Comment by joyously** [comment]
> I don't think the headers from the unified diff format add much. 

Are you thinking of humans or algorithms?
Same question applies regarding using snapshot and/or diff.

**Reply by martinvonz** [reply]
I was thinking of humans, but I don't see any benefit for algorithms either. Let me know if you do, of course.
